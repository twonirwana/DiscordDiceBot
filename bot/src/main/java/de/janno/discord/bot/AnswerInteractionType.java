package de.janno.discord.bot;

public enum AnswerInteractionType {
    none,
    hidden,
    reroll
}
