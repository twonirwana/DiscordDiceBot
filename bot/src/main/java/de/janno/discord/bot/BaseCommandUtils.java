package de.janno.discord.bot;

import de.janno.discord.bot.persistance.Mapper;
import de.janno.discord.bot.persistance.MessageDataDTO;
import de.janno.discord.bot.persistance.PersistenceManager;
import lombok.NonNull;

import javax.annotation.Nullable;
import java.util.UUID;

public class BaseCommandUtils {

    public static MessageDataDTO createEmptyMessageData(@NonNull UUID configUUID,
                                                        @Nullable Long guildId,
                                                        long channelId,
                                                        long messageId,
                                                        String commandId,
                                                        PersistenceManager persistenceManager) {
        MessageDataDTO messageDataDTO = new MessageDataDTO(configUUID, guildId, channelId, messageId, commandId, Mapper.NO_PERSISTED_STATE, null);
        //should not be needed but sometimes there is a retry ect and then there is already a state
        persistenceManager.deleteStateForMessage(channelId, messageId);
        persistenceManager.saveMessageData(messageDataDTO);
        return messageDataDTO;
    }
}
