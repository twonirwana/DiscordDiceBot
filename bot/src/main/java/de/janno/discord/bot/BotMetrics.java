package de.janno.discord.bot;

import com.google.common.base.Strings;
import io.micrometer.core.instrument.Tags;
import io.micrometer.core.instrument.Timer;
import io.micrometer.core.instrument.binder.jvm.*;
import io.micrometer.core.instrument.binder.logging.LogbackMetrics;
import io.micrometer.core.instrument.binder.system.ProcessorMetrics;
import io.micrometer.core.instrument.binder.system.UptimeMetrics;
import io.micrometer.prometheus.PrometheusConfig;
import io.micrometer.prometheus.PrometheusMeterRegistry;
import io.undertow.Handlers;
import io.undertow.Undertow;
import io.undertow.server.handlers.PathHandler;
import io.undertow.util.Headers;
import lombok.NonNull;

import java.time.Duration;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

public class BotMetrics {

    public final static String METRIC_PREFIX = "dice.";
    public final static String METRIC_BUTTON_PREFIX = "buttonEvent";
    public final static String METRIC_SLASH_PREFIX = "slashEvent";
    public final static String METRIC_SLASH_HELP_PREFIX = "slashHelpEvent";
    public final static String CONFIG_TAG = "config";
    public final static String COMMAND_TAG = "command";

    public static void init(String publishMetricsToUrl) {
        if (!Strings.isNullOrEmpty(publishMetricsToUrl)) {
            PrometheusMeterRegistry prometheusRegistry = new PrometheusMeterRegistry(PrometheusConfig.DEFAULT);
            io.micrometer.core.instrument.Metrics.addRegistry(prometheusRegistry);
            new UptimeMetrics().bindTo(globalRegistry);
            PathHandler handler = Handlers.path().addExactPath("prometheus", exchange ->
            {
                exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "text/plain");
                exchange.getResponseSender().send(prometheusRegistry.scrape());
            });
            Undertow server = Undertow.builder()
                    .addHttpListener(8080, publishMetricsToUrl)
                    .setHandler(handler).build();
            server.start();

            prometheusRegistry.config().commonTags("application", "DiscordDiceBot");
            new JvmMemoryMetrics().bindTo(globalRegistry);
            try(JvmGcMetrics jvmGcMetrics = new JvmGcMetrics()){
                jvmGcMetrics.bindTo(globalRegistry);
            }
            new ProcessorMetrics().bindTo(globalRegistry);
            new JvmThreadMetrics().bindTo(globalRegistry);
            try(LogbackMetrics logbackMetrics = new LogbackMetrics()){
                logbackMetrics.bindTo(globalRegistry);
            }
            new ClassLoaderMetrics().bindTo(globalRegistry);
            try(JvmHeapPressureMetrics jvmHeapPressureMetrics = new JvmHeapPressureMetrics()){
                jvmHeapPressureMetrics.bindTo(globalRegistry);
            }
            new JvmInfoMetrics().bindTo(globalRegistry);
        }
    }


    public static void incrementButtonMetricCounter(@NonNull String commandName, @NonNull String configString) {
        globalRegistry.counter(METRIC_PREFIX + METRIC_BUTTON_PREFIX, Tags.of(COMMAND_TAG, commandName, CONFIG_TAG, configString)).increment();
    }

    public static void incrementSlashStartMetricCounter(@NonNull String commandName, @NonNull String configString) {
        globalRegistry.counter(METRIC_PREFIX + METRIC_SLASH_PREFIX, Tags.of(COMMAND_TAG, commandName, CONFIG_TAG, configString)).increment();
    }

    public static void incrementSlashHelpMetricCounter(@NonNull String commandName) {
        globalRegistry.counter(METRIC_PREFIX + METRIC_SLASH_HELP_PREFIX, Tags.of(COMMAND_TAG, commandName)).increment();
    }
}
