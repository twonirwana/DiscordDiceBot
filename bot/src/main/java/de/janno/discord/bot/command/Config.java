package de.janno.discord.bot.command;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;

import java.io.Serializable;

/**
 * A configuration for a dice system. It will be created with the slash command and not modified afterwards.
 */
@EqualsAndHashCode
@Getter
@ToString
public class Config implements Serializable {

    private final Long answerTargetChannelId;

    @NonNull
    private final AnswerFormatType answerFormatType;

    @JsonCreator
    public Config(@JsonProperty("answerTargetChannelId") Long answerTargetChannelId,
                  @JsonProperty("answerFormatType") AnswerFormatType answerFormatType) {
        this.answerTargetChannelId = answerTargetChannelId;
        this.answerFormatType = answerFormatType == null ? AnswerFormatType.full : answerFormatType;
    }

    public String toShortString() {
        return String.format("[%s, %s]", getTargetChannelShortString(), answerFormatType);
    }

    protected String getTargetChannelShortString() {
        return answerTargetChannelId == null ? "local" : "target";
    }

}
