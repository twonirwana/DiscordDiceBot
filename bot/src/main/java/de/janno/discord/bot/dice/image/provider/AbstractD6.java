package de.janno.discord.bot.dice.image.provider;

import com.google.common.collect.ImmutableMap;
import lombok.NonNull;

import java.awt.image.BufferedImage;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

public abstract class AbstractD6 implements ImageProvider{

    private final Map<String, FileSidesDiceImageMap> color2DiceSideImageMap;

    public AbstractD6(String folderName) {
        this.color2DiceSideImageMap = getSupportedColors().stream()
                .collect(ImmutableMap.toImmutableMap(Function.identity(), c -> new FileSidesDiceImageMap(folderName + c, List.of(6))));
    }

    @Override
    public @NonNull List<BufferedImage> getImageFor(int totalDieSides, int shownDieSide, String color) {
        final String validatedColor;
        if (color == null || !getSupportedColors().contains(color)) {
            validatedColor = getDefaultColor();
        } else {
            validatedColor = color;
        }

        FileSidesDiceImageMap fileSidesDiceImageMap = color2DiceSideImageMap.get(validatedColor);
        return Optional.ofNullable(fileSidesDiceImageMap.getDiceImageMap().get(totalDieSides))
                .map(m -> m.get(shownDieSide))
                .map(List::of).orElse(List.of());
    }
}
