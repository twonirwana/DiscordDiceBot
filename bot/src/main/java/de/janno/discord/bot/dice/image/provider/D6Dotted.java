package de.janno.discord.bot.dice.image.provider;

import com.google.common.collect.ImmutableMap;
import lombok.NonNull;

import java.awt.image.BufferedImage;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

public class D6Dotted implements ImageProvider {

    public static final String WHITE = "white";
    public static final String BLACK_AND_GOLD = "black_and_gold";
    private static final List<String> SUPPORTED_COLORS = List.of(WHITE, BLACK_AND_GOLD);
    private final Map<String, FileSidesDiceImageMap> color2DiceSideImageMap;


    public D6Dotted() {
        this.color2DiceSideImageMap = SUPPORTED_COLORS.stream()
                .collect(ImmutableMap.toImmutableMap(Function.identity(), c -> new FileSidesDiceImageMap("d6_" + c, List.of(6))));
    }

    @Override
    public int getDieHighAndWide() {
        return 50;
    }

    @Override
    public @NonNull String getDefaultColor() {
        return WHITE;
    }

    @Override
    public @NonNull List<String> getSupportedColors() {
        return SUPPORTED_COLORS;
    }

    @Override
    public @NonNull List<BufferedImage> getImageFor(Integer totalDieSides, Integer shownDieSide, String color) {
        if (totalDieSides == null || shownDieSide == null) {
            return List.of();
        }

        final String validatedColor;
        if (color == null || !getSupportedColors().contains(color)) {
            validatedColor = getDefaultColor();
        } else {
            validatedColor = color;
        }

        FileSidesDiceImageMap fileSidesDiceImageMap = color2DiceSideImageMap.get(validatedColor);
        return Optional.ofNullable(fileSidesDiceImageMap.getDiceImageMap().get(totalDieSides))
                .map(m -> m.get(shownDieSide))
                .map(List::of).orElse(List.of());
    }
}
