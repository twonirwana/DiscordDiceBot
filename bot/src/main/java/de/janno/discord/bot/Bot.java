package de.janno.discord.bot;


import com.google.common.collect.ImmutableList;
import de.janno.discord.bot.command.*;
import de.janno.discord.bot.command.countSuccesses.CountSuccessesCommand;
import de.janno.discord.bot.command.customDice.CustomDiceCommand;
import de.janno.discord.bot.command.customParameter.CustomParameterCommand;
import de.janno.discord.bot.command.fate.FateCommand;
import de.janno.discord.bot.command.holdReroll.HoldRerollCommand;
import de.janno.discord.bot.command.poolTarget.PoolTargetCommand;
import de.janno.discord.bot.command.sumCustomSet.SumCustomSetCommand;
import de.janno.discord.bot.command.sumDiceSet.SumDiceSetCommand;
import de.janno.discord.bot.persistance.MessageDataDAO;
import de.janno.discord.bot.persistance.MessageDataDAOImpl;
import de.janno.discord.connector.DiscordConnectorImpl;

public class Bot {
    public static void main(final String[] args) throws Exception {
        final String token = args[0];
        final boolean disableCommandUpdate = Boolean.parseBoolean(args[1]);
        final String publishMetricsToUrl = args[2];
        BotMetrics.init(publishMetricsToUrl);

        final String h2Url;
        if (args.length >= 4) {
            h2Url = args[3];
        } else {
            h2Url = "jdbc:h2:file:./persistence/dice_config;AUTO_SERVER=TRUE;DB_CLOSE_DELAY=10";
        }
        final String h2User;
        if (args.length >= 5) {
            h2User = args[4];
        } else {
            h2User = null;
        }
        final String h2Password;
        if (args.length >= 6) {
            h2Password = args[5];
        } else {
            h2Password = null;
        }

        MessageDataDAO messageDataDAO = new MessageDataDAOImpl(h2Url, h2User, h2Password);

        DiscordConnectorImpl.createAndStart(token, disableCommandUpdate, ImmutableList.of(
                        new CountSuccessesCommand(messageDataDAO),
                        new CustomDiceCommand(messageDataDAO),
                        new FateCommand(messageDataDAO),
                        new DirectRollCommand(),
                        new SumDiceSetCommand(messageDataDAO),
                        new SumCustomSetCommand(messageDataDAO),
                        new HoldRerollCommand(messageDataDAO),
                        new PoolTargetCommand(messageDataDAO),
                        new CustomParameterCommand(messageDataDAO),
                        new WelcomeCommand(messageDataDAO),
                        new ClearCommand(messageDataDAO),
                        new BetaRollCommand(),
                        new HelpCommand()
                ),
                new WelcomeCommand(messageDataDAO).getWelcomeMessage());
    }

}
