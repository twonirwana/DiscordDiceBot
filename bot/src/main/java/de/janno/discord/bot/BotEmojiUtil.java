package de.janno.discord.bot;

import io.avaje.config.Config;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import net.fellbaum.jemoji.EmojiManager;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class BotEmojiUtil {

    private static final Pattern START_WITH_EMOJI_PATTERN = Pattern.compile("^<a?:([a-zA-Z0-9_]+):([0-9]+)>");

    public static String replaceEmojiKeyWithEmoji(String input) {
        if (input == null) {
            return null;
        }
        for (EmojiKey emojiKey : EmojiKey.values()) {
            String discordEmoji = Config.getOptional(emojiKey.getKey()).orElse("");
            input = input.replace(emojiKey.getKey(), discordEmoji);
        }
        return input;
    }

    public static String toDiscordString(EmojiKey emojiKey) {
        if (emojiKey == null) {
            return null;
        }
        return Config.getOptional(emojiKey.getKey()).orElse(null);
    }

    private static String getLeadingUnicodeEmoji(String in) {
        if (EmojiManager.containsEmoji(in)
                && in.startsWith(EmojiManager.extractEmojisInOrder(in).getFirst().getEmoji())) {
            return EmojiManager.extractEmojisInOrder(in).getFirst().getEmoji();
        }
        return null;
    }

    private static String getLeadingDiscordEmoji(String in) {
        Matcher emojiMatcher = START_WITH_EMOJI_PATTERN.matcher(in);
        boolean found = emojiMatcher.find();
        if (found) {
            return emojiMatcher.group();
        }
        return null;
    }

    public static LabelAndEmoji splitLabel(final @NonNull String label) {
        final String emoji;
        final String labelWithoutLeadingEmoji;
        final String cleanLabel = label.trim();
        final String unicodeEmoji = BotEmojiUtil.getLeadingUnicodeEmoji(cleanLabel);
        if (unicodeEmoji != null) {
            emoji = unicodeEmoji;
            labelWithoutLeadingEmoji = label.substring(unicodeEmoji.length());
            return new LabelAndEmoji(labelWithoutLeadingEmoji, emoji);
        }
        final String discordEmoji = BotEmojiUtil.getLeadingDiscordEmoji(cleanLabel);
        if (discordEmoji != null) {
            emoji = discordEmoji;
            //can produce an empty String
            labelWithoutLeadingEmoji = label.substring(discordEmoji.length());
            return new LabelAndEmoji(labelWithoutLeadingEmoji, emoji);

        }
        return new LabelAndEmoji(label, null);
    }

    @Getter
    @RequiredArgsConstructor
    public enum EmojiKey {
        BACK_EMOJI_KEY("emoji.back"),
        ROLL_EMOJI_KEY("emoji.roll"),
        CANCEL_EMOJI_KEY("emoji.cancel"),
        FATE_EMOJI_KEY("emoji.fate"),
        D20S20_RDD_EMOJI_KEY("emoji.d20s20_rdd"),
        D6S6_RDD_EMOJI_KEY("emoji.d6s6_rdd"),
        COIN_EMOJI_KEY("emoji.coinflip"),
        D6S6_BLACK_GOLD_EMOJI_KEY("emoji.d6s6_black_gold"),
        D10S10_RDD_EMOJI_KEY("emoji.d10s10_rdd"),
        D10S10_RED_EMOJI_KEY("emoji.d10s10_red"),
        D10S10_BLUE_EMOJI_KEY("emoji.d10s10_blue"),
        CALCULATOR_EMOJI_KEY("emoji.calculator");

        private final String key;

    }

    public record LabelAndEmoji(String labelWithoutLeadingEmoji, String emoji) {
    }
}
