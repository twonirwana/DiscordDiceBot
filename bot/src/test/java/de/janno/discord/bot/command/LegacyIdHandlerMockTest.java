package de.janno.discord.bot.command;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.ButtonEventAdaptorMock;
import io.avaje.config.Config;
import org.apache.commons.io.FileUtils;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.io.File;
import java.io.IOException;
import java.util.concurrent.atomic.AtomicLong;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SnapshotExtension.class)
public class LegacyIdHandlerMockTest {
    private Expect expect;

    @BeforeEach
    void setup() throws IOException {
        File cacheDirectory = new File("imageCache/");
        if (cacheDirectory.exists()) {
            FileUtils.cleanDirectory(cacheDirectory);
        }
    }

    @AfterEach
    void cleanUp() throws IOException {
        File cacheDirectory = new File("imageCache/");
        if (cacheDirectory.exists()) {
            FileUtils.cleanDirectory(cacheDirectory);
        }
        Config.setProperty("diceEvaluator.cacheSize", "0");
    }

    @ParameterizedTest
    @CsvSource({
            "count_successes, true",
            "fate, true",
            "hold_reroll, true",
            "pool_target, true",
            "sum_dice_set, true",
            "custom_dice, false",
    })
    void handleId(String commandId, boolean expectMatchingId) {
        LegacyIdHandler underTest = new LegacyIdHandler();
        ButtonEventAdaptorMock buttonEvent = new ButtonEventAdaptorMock(commandId, "button_1", new AtomicLong(0));

        boolean matchingId = underTest.matchingComponentCustomId(buttonEvent.getCustomId());
        assertThat(matchingId).isEqualTo(expectMatchingId);

        if (matchingId) {
            underTest.handleComponentInteractEvent(buttonEvent).block();
            expect.scenario(commandId).toMatchSnapshot(buttonEvent.getSortedActions());
        }
    }
}
