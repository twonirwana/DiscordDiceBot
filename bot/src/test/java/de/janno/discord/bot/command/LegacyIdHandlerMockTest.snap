de.janno.discord.bot.command.LegacyIdHandlerMockTest.handleId[count_successes]=[
  [
    "reply: Command is not supported anymore. Please create a new button message with the current commands."
  ]
]


de.janno.discord.bot.command.LegacyIdHandlerMockTest.handleId[fate]=[
  [
    "reply: Command is not supported anymore. Please create a new button message with the current commands."
  ]
]


de.janno.discord.bot.command.LegacyIdHandlerMockTest.handleId[hold_reroll]=[
  [
    "reply: Command is not supported anymore. Please create a new button message with the current commands."
  ]
]


de.janno.discord.bot.command.LegacyIdHandlerMockTest.handleId[pool_target]=[
  [
    "reply: Command is not supported anymore. Please create a new button message with the current commands."
  ]
]


de.janno.discord.bot.command.LegacyIdHandlerMockTest.handleId[sum_dice_set]=[
  [
    "reply: Command is not supported anymore. Please create a new button message with the current commands."
  ]
]